dandiapi/api/tasks/__init__.py

```diff
@@ -184,3 +185,45 @@ def unembargo_dandiset(dandiset_id: int):
     # Set access on dandiset
     dandiset.embargo_status = Dandiset.EmbargoStatus.OPEN
     dandiset.save(update_fields=['embargo_status'])
+
+
+@shared_task
+@atomic
+def publish_task(version_id: int):
+    old_version: Version = Version.objects.get(id=version_id)
+    new_version: Version = old_version.publish_version
+
+    new_version.doi = doi.create_doi(new_version)
+    new_version.save()
+
+    # Bulk create the join table rows to optimize linking assets to new_version
+    AssetVersions = Version.assets.through
+
+    # Add a new many-to-many association directly to any already published assets
+    already_published_assets = old_version.assets.filter(published=True)
+    AssetVersions.objects.bulk_create(
+        [
+            AssetVersions(asset_id=asset['id'], version_id=new_version.id)
+            for asset in already_published_assets.values('id')
+        ]
+    )
+
+    # Publish any draft assets
+    draft_assets = old_version.assets.filter(published=False).all()
+    for draft_asset in draft_assets:
+        draft_asset.publish()
+    Asset.objects.bulk_update(draft_assets, ['metadata', 'published'])
+
+    AssetVersions.objects.bulk_create(
+        [AssetVersions(asset_id=asset.id, version_id=new_version.id) for asset in draft_assets]
+    )
```

Comment:
All of these loops that iterate over every asset in a dandiset are probably going to lead to memory issues when publishing dandisets with large numbers of assets. These are copied 1-1 from the original publishing endpoint code, so I'm not going to attempt to fix them here; I'll do some memory profiling and make a follow up PR once I come up with a solution to this